action/protocol/staking/candidate_center.go

```diff
@@ -7,143 +7,432 @@
 package staking
 
 import (
+	"sync"
+
 	"github.com/iotexproject/iotex-address/address"
 )
 
 type (
-	// CandidateCenter is a struct to manage the candidates
-	CandidateCenter struct {
+	// CandidateControl control the candidate center
+	CandidateControl interface {
+		CandidateView
+		Base() CandidateControl
+		Delta() CandidateList
+		SetDelta(CandidateList) error
+		Commit(CandidateList) error
+	}
+
+	// CandidateView provides access to all candidates
+	CandidateView interface {
+		Size() int
+		All() CandidateList
+		ContainsName(string) bool
+		ContainsOwner(address.Address) bool
+		ContainsOperator(address.Address) bool
+		ContainsSelfStakingBucket(uint64) bool
+		GetByName(string) *Candidate
+		GetByOwner(address.Address) *Candidate
+		GetBySelfStakingIndex(uint64) *Candidate
+		Upsert(*Candidate) error
+		collision(*Candidate) error
+	}
+
+	// candChange captures the change to candidates
+	candChange struct {
+		dirty map[string]*Candidate
+	}
+
+	// candBase is the confirmed base state
+	candBase struct {
+		lock             sync.RWMutex
 		nameMap          map[string]*Candidate
 		ownerMap         map[string]*Candidate
 		operatorMap      map[string]*Candidate
 		selfStkBucketMap map[uint64]*Candidate
 	}
+
+	// CandidateCenter is a struct to manage the candidates
+	CandidateCenter struct {
+		*candBase
+		size   int
+		change CandidateView
+	}
 )
 
+// NewCandidateView creates an instance of CandidateView
+func NewCandidateView(l CandidateList) (CandidateView, error) {
+	cv := candChange{
+		dirty: make(map[string]*Candidate),
+	}
+
+	for _, d := range l {
+		if err := d.Validate(); err != nil {
+			return nil, err
+		}
+		cv.dirty[d.Owner.String()] = d
+	}
+	return &cv, nil
+}
+
 // NewCandidateCenter creates an instance of CandidateCenter
-func NewCandidateCenter() *CandidateCenter {
+func NewCandidateCenter() CandidateControl {
 	return &CandidateCenter{
-		nameMap:          make(map[string]*Candidate),
-		ownerMap:         make(map[string]*Candidate),
-		operatorMap:      make(map[string]*Candidate),
-		selfStkBucketMap: make(map[uint64]*Candidate),
+		candBase: &candBase{
+			nameMap:          make(map[string]*Candidate),
+			ownerMap:         make(map[string]*Candidate),
+			operatorMap:      make(map[string]*Candidate),
+			selfStkBucketMap: make(map[uint64]*Candidate),
+		},
+		change: newCandChange(),
 	}
 }
 
 // Size returns number of candidates
-func (m CandidateCenter) Size() int {
-	return len(m.nameMap)
+func (m *CandidateCenter) Size() int {
+	return m.size
 }
 
 // All returns all candidates in candidate center
-func (m CandidateCenter) All() (CandidateList, error) {
-	list := make(CandidateList, 0, m.Size())
-	for _, c := range m.nameMap {
-		list = append(list, c)
+func (m *CandidateCenter) All() CandidateList {
+	list := m.change.All()
+	if list == nil {
+		list = make(CandidateList, 0, len(m.ownerMap))
+	}
+
+	m.lock.RLock()
+	defer m.lock.RUnlock()
```

Comment:
This is an example we shouldn't do that, lock is in the middle, and it is so hard to understand or remember that lock is not ownerMap.

Reply:
fixed in latest push